src/misc/umap.c

```diff
@@ -0,0 +1,846 @@
+ *
+ * UMAP is a mostly used to embed high-dimensional vectors in a low-dimensional space
+ * (most commonly by far, 2D). The algorithm is probabilistic and introduces
+ * nonlinearities, unlike e.g. PCA and similar to T-distributed Stochastic Neighbor
+ * Embedding (t-SNE). Nonlinearity helps "cluster" very similar vectors together without
+ * imposing a global geometry on the embedded space (e.g. a rigid rotation + compression
+ * in PCA).
+ *
+ * However, UMAP uses a graph with distances associated to the edges as a key
+ * intermediate representation of the high-dimensional space, so it is also useful as
+ * a general graph layouting algorithm, hence its inclusion in igraph.
+ *
+ * Importantly, the edge-associated distances are derived from a similarity metric
+ * between the high-dimensional vectors, often Pearson correlation:
+ *
+ * corr(v1, v2) = v1 x v2 / [ sqrt(v1 x v1) * sqrt(v2 x v2) ]
+ *
+ * In this case, the associated distance is usually defined as:
+ *
+ * d(v1, v2) = 1 - corr(v1, v2)
+ *
+ * This implementation can also work with unweighted similarity graphs, in which case
+ * the distance parameter should be a null pointer and all edges beget a similarity
+ * score of 1 (a distance of 0).
+ *
+ * While all similarity graphs are theoretically embeddable, UMAP's stochastic gradient
+ * descent approach really shines when the graph is sparse. In practice, most people
+ * feed a k-nearest neighbor (either computed exactly or approximated) similarity graph
+ * with some additional cutoff to exclude "quasi-neighbors" that lie beyond a certain
+ * distance (e.g. correlation less than 0.2).
+ *
+ * Therefore, if you are trying to use this function to embed high-dimensional vectors,
+ * the steps are:
+ *
+ * 1. Compute a sparse similarity graph (either exact or approximate) from your vectors,
+ *    weighted or unweighted. If unsure, compute a knn.
+ * 2. If you keep the weights, convert them into distances or store them as a "weight"
+ *    edge attribute and use a null pointer for the distances. If using similarity
+ *    weights instead of distances, make sure they do not exceed 1.
+ * 3. Feed the graph (and distances, if you have them) into this function.
+ *
+ * Note: Step 1 above involves deciding if two high-dimensional vectors "look similar"
+ *       which, because of the curse of dimensionality, is in many cases a highly
+ *       subjective and potentially controversial operation: thread with care and at
+ *       your own risk. Two high-dimensional vectors might look similar or extremely
+ *       different depending on the point of view/angle, and there are a lot of
+ *       viewpoints when the dimensionality ramps up.
+ *
+ * </para><para>
+ * References:
+ *
+ * </para><para>
+ * Leland McInnes, John Healy, and James Melville, arXiv:1802.03426v3
+ *
+ * \param graph Pointer to the similarity graph to find a layout for (i.e. to embed). Weights in
+ *   this graph object are currently ignored.
```

Comment:
What do you mean by "weights are ignored"? Are you referring to edge attributes? If so, this comment should be removed. All igraph functions that use weights take it as a separate vector parameter instead of relying on the presence of an attribute with a hardcoded name. Generally, igraph functions should be usable without having the attribute handler enabled.